connections/muxer-sel-in-sec-handshake.md

```diff
@@ -0,0 +1,220 @@
+tunnel
+3. The multistream-selection protocol then will run again for muxer selection.
+4. Connection then is upgraded to a capable connection by the selected muxer.
+
+
+## Improved muxer selection
+
+The security protocol's ability of supporting higher level abstract protocol
+negotiation (for example, TLS's support of NextProtos, and Noise's support of
+Early Data) makes it possible to collapse the step 2 and step 3 in the
+previous section into one step. Muxer selection can be performed as part of
+the security protocol handshake, thus there is no need to perform another
+mutistream-selection negotiation for muxer selection.
+
+In order to achieve the above stated goal, each candidate muxer will be
+represented by a protocol name/code, and the candidate muxers are supplied to
+the security protocol's handshake process as a list of protocol names.
+
+If the client and server agree upon the common muxer to be used, then the
+result of the muxer selection is a muxer code represented by the selected
+protocol name. If no agreement is reached upon by the client and server
+then an empty muxer code is returned and the connection upgrade process
+MUST fall back to the multistream-selection protocol to negotiate the muxer.
+
+
+### Muxer selection over TLS
+
+When the security protocol selected by the upgrader is TLS, the [ALPN]
+extesion of TLS handshake is used to select the muxer.
+
+   With ALPN, the client sends the list of supported application
+   protocols as part of the TLS ClientHello message.  The server chooses
+   a protocol and sends the selected protocol as part of the TLS
+   ServerHello message.  The application protocol negotiation can thus
+   be accomplished within the TLS handshake, without adding network
+   round-trips, and allows the server to associate a different
+   certificate with each application protocol, if desired.
+
+For the purpose of muxer selection, the types of muxers are coded as protocol
+names in the form of a list of strings, and inserted in the ALPN "NextProtos"
+field. An example list as following:
+
+    ["yamux/1.0.0", "mplux", "libp2p"]
+
+The NextProtos list is ordered by preference, with the most prefered muxer at
+the beginning. The "libp2p" protocol code MUST always be the last item in the
+ALPN NextProtos list.
+
+The server chooses the supported protocol by going through its prefered
+protocol list and searchs if the protocol is supported by the client too. if no
+mutually supported protcol is found the TLS handshake will fail.
+
+If the selected NextProto is "libp2p" then the muxer selection process returns
+an empty result, and the multistream-selection protocol MUST be run to negotiate
+the muxer.
+
+(TBD: in some cases early abortion is desireable)
+
+
+### Muxer selection over Noise
+
+The libp2p Noise implementation allows the Noise handshake process to carry
+early data. [Noise-Early-Data] is carried in the second and third message of
```

Comment:
We need to specify which one exactly.

Reply:
sure I can add more specifics on handshaking/early data interaction.

Reply:
There'd be an advantage to using the first and the second handshake message instead of the second and the third: After receiving the list of muxers from the client, the server would already know which muxer was chosen, and could already initialize that muxer and start opening streams.
This would allow the server to optimistically (at this point, the client is not yet authenticated, so we can't trust its peer ID) send some data.

For example, I could see us opening an identify stream to push our list of supported protocols at this point. We don't really care if this list is captured by a MITM, but it would allow the client to proceed faster after the handshake.

What do you think?

Reply:
does the data actually get sent over the wire? I thought the write would be blocked until the noise handshake finishes.

Reply:
Given TLS already has muxer info in the plain text for the first TLS handshake message, and the fact muxer info is not sensitive data, I don't have objections to transmit the muxer info in the first and second Noise handsahke messages.

Reply:
> the server would already know which muxer was chosen, and could already initialize that muxer and start opening streams.
This would allow the server to optimistically (at this point, the client is not yet authenticated, so we can't trust its peer ID) send some data.

I don't think we could do this if we are encrypting the connection with the key derived from the noise handshake. The data that's sent after the handshake completes must be derived from all the DH operations. If the server optimistically starts sending data (e.g. initializing the muxer and opening streams) it will be encrypted with the wrong key after the noise handshake completes. The server needs to wait for the 3rd handshake message (in XX) for it to know what key to use to encrypt the data over the wire.

Reply:
Per the discussion on #454, we decided to use first and second messages to send early data, so resolving this conversation thread.